TLM/TLM/Manager/Impl/LaneConnectionManagerData/ConnectionData.cs

```diff
@@ -0,0 +1,159 @@
+namespace TrafficManager.Manager.Impl.LaneConnectionManagerData {
+    using CSUtil.Commons;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using TrafficManager.Util;
+    using TrafficManager.Util.Extensions;
+    using static TrafficManager.Util.Shortcuts;
+    internal struct LaneConnectionData {
+        public uint LaneId;
+        // TODO [issue 354]: add track/car connection type.
+
+        public LaneConnectionData(uint laneId) {
+            LaneId = laneId;
+        }
+
+        public override string ToString() => $"TargetConnectionData({LaneId})";
+    }
+
+    /// <summary>
+    /// All calls to this class assumes lanes are valid
+    /// </summary>
+    internal class ConnectionData {
+        private SourceLaneConnectionData[] connections_;
+
+        public ConnectionData() {
+            connections_ = new SourceLaneConnectionData[NetManager.instance.m_lanes.m_size];
+        }
+
+        private struct SourceLaneConnectionData {
+            public LaneConnectionData[] StartConnections;
+            public LaneConnectionData[] EndConnections;
+
+            public IEnumerable<LaneConnectionData[]> BothConnections() {
+                yield return StartConnections;
+                yield return EndConnections;
+            }
+        }
+
+        internal ref LaneConnectionData[] GetConnections(uint laneId, ushort nodeId) {
+            return ref GetConnections(laneId, laneId.ToLane().IsStartNode(nodeId));
+        }
+
+        internal ref LaneConnectionData[] GetConnections(uint laneId, bool startNode) {
+            if(startNode)
+                return ref connections_[laneId].StartConnections;
+            else
+                return ref connections_[laneId].EndConnections;
+        }
+
+        internal bool IsConnectedTo(uint sourceLaneId, uint targetLaneId, ushort nodeId) =>
+            IsConnectedTo(sourceLaneId, targetLaneId, sourceLaneId.ToLane().IsStartNode(nodeId));
+
+        /// <param name="sourceStartNode">start node for the segment of the source lane</param>
+        internal bool IsConnectedTo(uint sourceLaneId, uint targetLaneId, bool sourceStartNode) {
+            var targets = GetConnections(sourceLaneId, sourceStartNode);
```

Comment:
by pointer you mean array?
what do you mean by unsafe? I am not using unsafe block. why is this unsafe? I don't see any problem.

Reply:
compilation result:
![image](https://user-images.githubusercontent.com/19638970/157723598-85f2da4d-9232-44e6-afd9-a5c9067706ea.png)

It's because `GetConnections()` returns `ref` explicitly and you don't store it as ref but as value

Reply:
Oh I see. I didn't know that?

I don't store it as ref because I don't need to. Should I store it as ref anyway to avoid the unsafe block?
what do you think is the solution?

Reply:
also this:
![image](https://user-images.githubusercontent.com/19638970/157730888-35ee2e77-fa3e-4771-a06b-5d50697fb8b2.png)